### PR TITLE
interpret: don't rely on ScalarPair for overflowed arithmetic

### PR DESCRIPTION
**Pull request by RalfJung**
This is for https://github.com/rust-lang/rust/pull/97861.
Cc @eddyb 

I would like to avoid making this depend on `dest.layout.abi` to avoid a branch that we are not usually covering both sides of. Though OTOH this seems like fairly straight-forward code. But let's benchmark this option first to see how bad that extra `force_allocation` really is.

### CODE REVIEW

**Comment by RalfJung** [comment]
That is strange, is this making const-prop worse?

**Comment by bjorn3** [comment]
Cg_clif makes the same assumption with respect to the `ScalarPair` ABI being used for `CheckedBinOp`.

**Comment by oli-obk** [comment]
Oh yea this affects const prop. Const prop likely doesn't handle locals backed by an alloc properly and optimizes/lints less

**Comment by RalfJung** [comment]
Aand I guess CI is using a debug build. Argh.
Is there the equivalent of `// ignore-debug` for run-make tests?

**Comment by RalfJung** [comment]
Well, this should fix CI. The only problem is that now the new code path is fully untested...

**Comment by lcnr** [comment]
i am fine with that

r=me unless you want to try something else

File: compiler/rustc_const_eval/src/interpret/operator.rs [hunk 1]
```diff
@@ -25,8 +26,21 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
         );
-        let val = Immediate::ScalarPair(val.into(), Scalar::from_bool(overflowed).into());
-        self.write_immediate(val, dest)
+        if let Abi::ScalarPair(..) = dest.layout.abi {
+            // We can use the optimized path and avoid `place_field` (which might do
+            // `force_allocation`).
+            let pair = Immediate::ScalarPair(val.into(), Scalar::from_bool(overflowed).into());
+            self.write_immediate(pair, dest)?;
+        } else {
+            // With randomized layout, `(int, bool)` might cease to be a `ScalarPair`, so we have to
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/interpret/operator.rs:35]: maybe explicitly assert that we're using `-Zrandomize-layout` (or however the flag is called) here?

**Comment by Kixiron** [comment]
Sure, I can totally try to test layout randomization in a later PR

**Pull request merged**